index/segment/postings_list_pool.go

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package segment
+
+import (
+	"github.com/m3db/m3x/pool"
+)
+
+type postingsListPool struct {
+	pool pool.ObjectPool
+}
+
+// PostingsListAllocateFn returns a new PostingsList.
+type PostingsListAllocateFn func() PostingsList
+
+// NewPostingsListPool returns a new PostingsListPool.
+func NewPostingsListPool(
+	opts pool.ObjectPoolOptions,
+	allocator PostingsListAllocateFn,
+) PostingsListPool {
+	p := &postingsListPool{
+		pool: pool.NewObjectPool(opts),
+	}
+	p.pool.Init(func() interface{} {
+		return allocator()
```

Comment:
Do you want to `pl.Reset()` before returning or should new alloc always have same state as a reset postingslist?

Reply:
yep, assuming the allocator returns a usable postings list to start. We reset on Put to ensure the objects in the pool are usable.

Reply:
Usually consolidating the "clean state" to a single method is ideal.